src/repr/type_core.ml

```diff
@@ -33,6 +33,23 @@ module Json = struct
     | [] -> Jsonm.decode e.d
 end
 
+module Encode_json = Attribute.Make1 (struct
+  type 'a t = Jsonm.encoder -> 'a -> unit
+
+  let name = "encode_json"
+end)
+
+module Decode_json = Attribute.Make1 (struct
+  type 'a t = json_decoder -> ('a, [ `Msg of string ]) result
+
+  let name = "decode_json"
+end)
+
+let annotate t ~add ~data =
+  match t with
+  | Attributes t -> Attributes { t with attrs = add data t.attrs }
+  | t -> Attributes { attrs = add data Attribute.Map.empty; attr_type = t }
```

Comment:
The main advantage is performance: O(log n) find and O(1) ignore, rather than O(n) for both if one needs to traverse to the end of the linked list. I'm not so sure it's simpler either, having a list of K-V pairs implemented via nesting sounds like just re-implementing a map to me.